### ISSUE TITLE
Model not learning objectness

### ISSUE THREAD

**Issue by meetyogi98** [ultralytics/yolov5#8577]
I trained the model with some data and got descent graphs As below:
![image](https://user-images.githubusercontent.com/33971561/178980935-69a24734-1add-4d25-b22f-2d5f814153ed.png)

After that, I change the whole data. Current data is different from the previous one the only difference is that now objects are smaller in size compared to the previous one and the size of images is also small. The current task is also simple in comparison to the previous one. I trained the model with different resolutions like 608,896 and 1280. Now the graph of objections looks as below:
![image](https://user-images.githubusercontent.com/33971561/178981834-c055361a-240c-4b47-8691-f06a4554570e.png)

Due these results are very bad from the previous one. There is quite a big jump in the FN of objects.

I want to know what might be the reasons for this that cause such changes in the results of the model.

**Comment by meetyogi98** [comment]
I have also edited below code:
`def box_candidates(box1, box2, wh_thr=2, ar_thr=20, area_thr=0.2):  # box1(4,n), box2(4,n)
    # Compute candidate boxes: box1 before augment, box2 after augment, wh_thr (pixels), aspect_ratio_thr, area_ratio    
    w1, h1 = box1[2] - box1[0], box1[3] - box1[1]
    w2, h2 = box2[2] - box2[0], box2[3] - box2[1]
    ar = np.maximum(w2 / (h2 + 1e-16), h2 / (w2 + 1e-16))  # aspect ratio
    return (w2 > wh_thr) & (h2 > wh_thr) & (w2 * h2 / (w1 * h1 + 1e-16) > area_thr) & (ar < ar_thr)  # candidates`

to this:
`def box_candidates(box1, box2, wh_thr=2, ar_thr=1280, area_thr=0.1):  # box1(4,n), box2(4,n)
    # Compute candidate boxes: box1 before augment, box2 after augment, wh_thr (pixels), aspect_ratio_thr, area_ratio
    w1, h1 = box1[2] - box1[0], box1[3] - box1[1]
    w2, h2 = box2[2] - box2[0], box2[3] - box2[1]
    ar = np.maximum(w2 / (h2 + 1e-16), h2 / (w2 + 1e-16))  # aspect ratio
    return (w2 > wh_thr) & (h2 > wh_thr) & (w2 * h2 / (w1 * h1 + 1e-16) > area_thr) & (ar < ar_thr)  # candidates`

**Comment by glenn-jocher** [comment]
@meetyogi98 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by meetyogi98** [comment]
Hi, @glenn-jocher as asked I'm sharing the required files below:
Image showing graphs:
![results](https://user-images.githubusercontent.com/33971561/179458855-b9b837e9-13b5-414c-892d-4fc86ca7da77.png)

Image showing labels statistics:
![labels](https://user-images.githubusercontent.com/33971561/179458938-230c9fb6-f0f8-40de-862e-c82032a84574.png)

Training augmentation images:
1)
![train_batch0](https://user-images.githubusercontent.com/33971561/179459071-e2d39bb4-fd43-4766-9745-e072e76802a0.jpg)
2)
![train_batch1](https://user-images.githubusercontent.com/33971561/179459081-ffad1cda-2485-4d75-897e-bfdb3c9161c2.jpg)
3)
![train_batch2](https://user-images.githubusercontent.com/33971561/179459085-cb68e051-5bbc-4912-9b9f-e9ccef6d21f1.jpg)

Prediction on the test:
Ground truth image
![test_batch0_gt](https://user-images.githubusercontent.com/33971561/179459194-bab1594c-244f-48c7-a7b1-bbadcf648f38.jpg)

Predicted image
![test_batch0_pred](https://user-images.githubusercontent.com/33971561/179459314-18149f5f-1ddf-4dd2-9bf7-ac72f7ea5ff2.jpg)

I really appreciate any help you can provide.

**Comment by glenn-jocher** [comment]
@meetyogi98 your test data doesn't much resemble your training data.
<img width="2029" alt="dataset" src="https://user-images.githubusercontent.com/26833433/179813461-ed4abfab-7116-4c9a-9eae-49e95f902881.png">

Also for text information you might want to disable left-right flips:
https://github.com/ultralytics/yolov5/blob/92e47b85d952274480c8c5efa5900e686241a96b/data/hyps/hyp.scratch-low.yaml#L31

**Comment by meetyogi98** [comment]
@glenn-jocher Agree with you about test data, there might be some issue with it. I recently tried turning off flip and then train even in that case the situation was the same. 

I trained with some different data which was somewhat similar to my current data in that case I was not facing such issues.
Let me share with you the graphs for the same:
![results](https://user-images.githubusercontent.com/33971561/179944857-0c3b3b68-2ae5-4eb2-80c6-c6117f1be350.png)


In the above case, I have got quite good results.
The above graph is of the data which is text-based in which I trained with default augmentation settings.

**Comment by glenn-jocher** [comment]
@meetyogi98 yes, above results look good. There's likely a dataset issue then, or perhaps the new data is simply harder to train on.

**Comment by meetyogi98** [comment]
Hi @glenn-jocher,
I did some experiments with the same data. I just reduced the high aspect ratio image to low aspect ratio images. As most of the images were above 10 aspect ratio I thought that might be the problem with data and could improve the model performance and can get the graphs to normal. It happened the same after the experiment.

Now I completely changed the data. In this data also there were high aspect ratio images which I converted to low aspect ratio images. This time the current data graphs were uncertain as also the results.

I'm sharing the required files of the training just have look at it and let me know if you find something that could solve the problem.

label.png
![labels](https://user-images.githubusercontent.com/33971561/184499316-278ee6c3-f7ec-4a05-bec0-ed82e874de4a.png)

result.png
![results](https://user-images.githubusercontent.com/33971561/184499285-c0c4b4cb-21d3-42eb-95d4-1d8690279609.png)

train_batch0
![train_batch0](https://user-images.githubusercontent.com/33971561/184499288-d43bd3c0-74a8-474f-9d36-71b52b438f97.jpg)

train_batch1
![train_batch1](https://user-images.githubusercontent.com/33971561/184499347-7db7109b-3296-4b0c-981e-ff97e20c88e7.jpg)

train_batch2
![train_batch2](https://user-images.githubusercontent.com/33971561/184499296-ae76c0c7-b530-4a38-8893-63c46483a465.jpg)

test_batch0_gt
![test_batch0_gt](https://user-images.githubusercontent.com/33971561/184499406-17f95984-aa16-44e8-9dad-db5c8a61954b.jpg)

test_batch0_pred
![test_batch0_pred](https://user-images.githubusercontent.com/33971561/184499300-e5cc6f56-2584-4073-83cb-be9bb355772a.jpg)

Thanks in advance.

**Comment by glenn-jocher** [comment]
@meetyogi98 from your results it looks like your YOLOv5 is quite outdated. You should update your code and retrain with the very latest code as it may help. Lastly it's possible that the problem you have is ill-suited to object detection. We try to cover as many use cases as possible, but not all are well suited. The rule of thumb is that is an average person can learn the objects in your images then YOLOv5 should be able to also.

To update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

**Comment by meetyogi98** [comment]
Thanks once again...

I have one more question in which I saw something different in two different datasets. I'm attaching the images for the same.

This is the first image in which in the test saved image we can see all annotations in the ground truth: 
![test_batch0_gt](https://user-images.githubusercontent.com/33971561/184577436-e3ac36bb-e2bc-4121-ba8a-9a73003741a7.jpg)

This is another image of a different dataset:
![test_batch0_gt](https://user-images.githubusercontent.com/33971561/184577361-95bf6f78-3568-49c2-a5b0-ef9b8f490715.jpg)

I checked the annotation of ground truth it is all good this only happens while the image is saved during training.

If this is the reason for the bad results I can improve the results.

Thanks in advance...

**Comment by glenn-jocher** [comment]
@meetyogi98 older versions of YOLOv5 would reject boxes with high aspect ratios (skinny or long boxes), so perhaps this is what you are seeing there, or perhaps your dataset is missing labels there. I would update to the very latest code and retrain, as the latest code accepts boxes with aspect ratios from 1/100 to 100/1 (w/h - h/w)

**Comment by meetyogi98** [comment]
But in the above image mentioned I have reduced the aspect ratio to make that work and doing that I have seen improvement in another dataset.

Yes, still I will once try with the latest code.
Thanks, @glenn-jocher that will help me a lot...

**Comment by glenn-jocher** [comment]
You're welcome, @meetyogi98. Feel free to reach out if you have any further questions or need assistance with the latest code. Good luck with your experiments!
